### PR TITLE
feat: Add Deep connector Instantly (Read)

### LINKED ISSUE #912: Add Instantly deep connector

### PR DESCRIPTION
**Pull request by Cobalt0s**
Closes #912 

# Description

There are few endpoints that perform listing. Those that do have varying endpoint paths.
Therefore, there is a hard coded mapping between object name (in plural) to the appropriate URL path.
Pagination is manually calculated as response is not a reliable source as it differes among endpoints.

# Testing
image

### CODE REVIEW

**Comment by Cobalt0s** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/923?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#923** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/923?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#922** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/922?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#914** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/914?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#915](https://github.com/amp-labs/connectors/pull/915) <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/915?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Cobalt0s and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: instantly/url.go
Comment (jlimatampersand) [inline instantly/url.go:9]: Same as the `WithClient` can we also remove the duplicate definition for `constructURL` and define in `common` package and reuse it?

File: instantly/parse.go
Comment (jlimatampersand) [inline instantly/parse.go:61]: same as other functionss, define in `common` and remove duplicate definitions.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]
For many part of the code seem to be redundant. I commented on some of them, but wherever applicable, lets move to common package and reduce duplicate code. 

You can do it just for this this connector for the moment, and apply to other connectors in follow up PR to separate the scopes.

File: instantly/params.go
Comment (Cobalt0s) [reply instantly/params.go:38]: This method so far looked identical, but there are more non-oauth2 connectors added which would change the arguments and the impl of the function.

We are using `paramsbuilder.Client` as a delegate. So each connector then selects one of `WithOauthClient` / `WithBasicClient` / `WithApiKeyHeaderClient` / `WithApiKeyQueryParamClient`.

Note: `paramsbuilder` has a dependency on `common` package. In case we need to move such methods somewhere it will be new package to avoid circular dependency. (I tried some refactoring while working on this PR and thats what I had run into).

I used delegation to have some consistency in the way parameters are defined. Since we are using Options pattern the repitition will be there.

File: instantly/read.go
Comment (Cobalt0s) [reply instantly/read.go:76]: Yes, dataLocation could be related to URL. I didn't know how to use it.
With link I try to call it url, from older code I kind of introduced link word... anyway I see that there is more semantic dfiference then I give to it.

I will change those, and pull the method out.

File: instantly/url.go
Comment (Cobalt0s) [reply instantly/url.go]: See: https://github.com/amp-labs/connectors/pull/972

File: instantly/parse.go
Comment (Cobalt0s) [reply instantly/parse.go]: Please, review: https://github.com/amp-labs/connectors/pull/971

File: instantly/read.go
Comment (jlimatampersand) [inline instantly/read.go:93]: let's use `url` instead of `link`

File: instantly/params.go
Comment (jlimatampersand) [reply instantly/params.go:38]: I looked into the code, I think we can define those functions once in. Looking at how much lines we are replicating for each connector, I think this design is wrong. We always need to think in scale. We will end up having 1000+ connectors. which means this will end up being hundreds of thousands of lines. Which is simply not scalable. I am not sure how we came up this design from the beginning, but if this is current way, we should stop and redesign this. 

We shouldn't declare `type Option = func(*parameters)` type for every connector. 

Option should be in a separate package so that those can be imported and reused, not redefined. 

Let's hold off on this until I can think through and redesign this.

File: instantly/parse.go
Comment (jlimatampersand) [inline instantly/parse.go:12]: again, this should not be defined for every connector and same for all other functions in this file. Maybe except for the `makeNextRecordsURL` since url are specific to the  provider.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]

File: instantly/read.go
Comment (Cobalt0s) [reply instantly/read.go]: Sorry, missed that one.
I did rename in other files too. This naming will stop.

File: instantly/parse.go
Comment (Cobalt0s) [reply instantly/parse.go:12]: I went further. You are right that there is repitition, in this case this method shouldn't even exist. Here is a PR: https://github.com/amp-labs/connectors/pull/991

File: instantly/read.go
Comment (Cobalt0s) [reply instantly/read.go]: It started to get messy for this PR, I split it: https://github.com/amp-labs/connectors/pull/993

File: instantly/params.go
Comment (Cobalt0s) [reply instantly/params.go:38]: First of all by design we require setting up connectors using  `With<MethodName>`. This means every connector must define this method. Also IDE completion makes it useful when say instantly.With... and it autocompletes potential arguments.

`Option`s golang alias is necessary for generic constructor. It provides typed parameter. Every connector must define its set of parameters. So there is nothing that can be refactored here. In other words type `Option = func(*connectorName.parameters)` is not common or shared in any way, only the visual is.

And about methods that live under `parse.go`. This file is responsible to understand response format of a provider. This again cannot be known universally and every connector stores array of items under different field name, everybody does pagination in their unique way. Maybe as more patterns become apparent we could add utility functions to shorten it.

The patterns that could evolve could be more configuration/data driven. But thats the information that makes the identity of that connector. I agree that adding connectors can be confusing especially to outsiders. But slowly these things are polished. I think what we did with `errors.go` can be done here (config driven).

**Review by jlimatampersand — APPROVED** [review]
Approving this PR for the time being. There were some comments that I had about having repetitive code for each connectors. I found refactoring this will take time and certainly this PR does not need to wait for it. But please do try to minimize the repetition for PRs later on. 

Currently we are adding 1000 lines per connectors, which will quickly become 1 million lines of code if we get to 1000 connectors. 

I will work on refactoring in next couple weeks to come up with better solution. It seems like this pattern has ben inherited from prior to your works. Thanks for working on this and other optimizations you are doing!

File: providers/instantly/parse.go
Comment (jlimatampersand) [inline providers/instantly/parse.go:18]: This and above function should be in common pacakge or utils package and imported to be used. Also `return jsonquery.New(node).ArraySize(dataLocation)` is being used twice for no reason and drawing performance back. 

Please resolve these before merging

**Review by jlimatampersand — CHANGES_REQUESTED** [review]
Please take out `makeGetRecords` to common package and reuse. Please create a follow up PR to implement this to other connectors as well.

File: providers/instantly/parse.go
Comment (jlimatampersand) [reply providers/instantly/parse.go:18]: I initially approved, but found there are some code `makeGetRecords` that is redefined for each connector and the function has nothing specific to the connector. Please move to common or utils and import to use.

Comment (Cobalt0s) [reply providers/instantly/parse.go:18]: I will create common utilities for this method.

Comment (Cobalt0s) [reply providers/instantly/parse.go]: Dependent PR: https://github.com/amp-labs/connectors/pull/1004

**Review by jlimatampersand — APPROVED** [review]

**Pull request merged**
